docs/reference/proto-definition.md

```diff
@@ -77,11 +77,11 @@ PROTO MyProto [
   field SFString                            name          "my proto"
   field SFColor{0 0 0, 0.5 0.5 0.5, 1 1 1}  color         0.5 0.5 0.5
   field SFNode                              physics       NULL
-  field MFNode{Solid{}, Pose{}}             extensionSlot []
+  field MFNode{Solid{}, Transform{}}        extensionSlot []
 ]
 ```
 
-In this example, the `color` field value can only be `0 0 0`, `0.5 0.5 0.5` or `1 1 1` and the `extensionSlot` field can only accept [Solid](../reference/solid.md) and [Pose](../reference/pose.md) nodes.
+In this example, the `color` field value can only be `0 0 0`, `0.5 0.5 0.5` or `1 1 1` and the `extensionSlot` field can only accept [Solid](../reference/solid.md), [Transform](../reference/transform.md) nodes, and nodes that inherit from them.
```

Comment:
This doesn't make sense any more, since `Solid` inherits from `Transform`, so we should write instead:
```js
  field MFNode{Transform{}}                    extensionSlot []
```
Also, I see a more fundamental problem here. In some case we may want to allow a `Transform` or a `Pose`, but not a Solid. The new system doesn't allow this any more, which may be problematic in some cases, like I want to have an optional shape on the top of my proto, which has no bounding object and no physics, thus I need a `Pose`, but not a `Solid`. How could the new system handle this?

Reply:
`Solid` inherits from `Pose` not `Transform`.

In the current model, the desired behavior could be softly enforced by declaring a `Pose` restriction and adding 
``` javascript
%<
// For SFNodes
assert(!(fields.field.boundingBox || fields.field.physics), "field must have no boundingObject or physics!");
// For MFNodes
assert(fields.field.every(node => !(node.boundingBox || node.physics)), "field must have no boundingObject or physics!");
>%
```
to the proto definition.

Alternatively, if you'd rather have the behavior to be supported by the proto syntax, we could define something like
``` js
field SFNode{=Pose{},=Transform{}} shape Pose{}
```
to restore the original behavior.

Reply:
> `Solid` inherits from `Pose` not `Transform`.

Oops. Yes, you are right. So the example fully makes sense. Maybe we could also explain that in this specific case we might have simply used `Pose` to encompass both `Transform` and `Solid`. However `Pose` would also allow for `Fluid` which may not be desirable.

> In the current model, the desired behavior could be softly enforced by declaring a `Pose` restriction and adding
> 
> ```js
> %<
> // For SFNodes
> assert(!(fields.field.boundingBox || fields.field.physics), "field must have no boundingObject or physics!");
> // For MFNodes
> assert(fields.field.every(node => !(node.boundingBox || node.physics)), "field must have no boundingObject or physics!");
> >%
> ```
> 
> to the proto definition.

Yes, that's indeed possible, although not very convenient.

> Alternatively, if you'd rather have the behavior to be supported by the proto syntax, we could define something like
> 
> ```js
> field SFNode{=Pose{},=Transform{}} shape Pose{}
> ```
> 
> to restore the original behavior.

Yes, my only remaining concern with this PR is that it breaks the backwards compatibility.

```js
field SFNode{=Pose{},=Transform{}} shape Pose{}
```
in the new version would be equivalent to:
```js
field SFNode{Pose{},Transform{}} shape Pose{}
```
in the old version.

If we want to maintain backwards compatibility, while still introducing your new feature, we could a syntax like:
```js
field SFNode{Solid*{},Transform*{}} shape Pose{}
```
to mean that and `Solid` or any descendant of `Solid` and `Transform` or any descendant of `Transform` is allowed. Though, I am not sure the `*` syntax is the best one (it doesn't look very nice to my eyes). Honestly, I prefer the syntax you propose with the `=` sign even if that breaks the backwards compatibility, because I believe that such a break will only have limited consequences. So, it you go ahead with the proposed `=` syntax, I am OK with that.

Reply:
In the latest batch of commits, I added the `=` case, but made the parser revert to the old behavior if the proto version was <R2024a. However, while typing this out, I tried out your suggestion with `+` rather than `*`, and it's starting to grow on me.
``` javascript
field SFNode{Solid+{},Transform+{}} shape Solid{}
```
(explicitly allows descendants)
vs
``` javascript
field SFNode{=Solid{},=Transform{}} shape Solid{}
```
(explicitly disallows descendants)

Ultimately, it might come down to a decision of which default behavior makes more sense. From an object-oriented/inheritance perspective, the `=` syntax makes more sense because, on some level, a descendant of a `Solid` *IS* a `Solid`. However, given that Webots seems to treat inheritance more as "distinct nodes that happen to share some similar features," (i.e. a decendant of a `Solid` just sometimes *LOOKS* like a `Solid`) the `+` syntax might make more sense for this PR.

I've been staring at this too long, so I think I'll sleep on it, but lmk if you have any thoughts.